lib/spack/llnl/util/filesystem.py

```diff
@@ -1734,83 +1739,78 @@ def find(root, files, recursive=True, max_depth: Optional[int] = None):
-        globs = [globs]
-    # Apply normcase to regular expressions and to the filenames:
-    # this respects case-sensitivity semantics of different OSes
-    # (e.g. file search is typically case-insensitive on Windows)
-    regexes = [re.compile(fnmatch.translate(os.path.normcase(x))) for x in globs]
-
-    # Note later calls to os.scandir etc. return abspaths if the
-    # input is absolute, see https://docs.python.org/3/library/os.html#os.DirEntry.path
-    root = os.path.abspath(root)
-
-    found_files = collections.defaultdict(list)
-
-    def _dir_id(stat_info):
-        # Note: on windows, st_ino is the file index and st_dev
-        # is the volume serial number. See
-        # https://github.com/python/cpython/blob/3.9/Python/fileutils.c
-        return (stat_info.st_ino, stat_info.st_dev)
+    If a directory has a name which matches an input pattern, it will not be included in the
+    results.
 
-    def _log_file_access_issue(e):
-        errno_name = errno.errorcode.get(e.errno, "UNKNOWN")
-        tty.debug(f"find must skip {dir_entry.path}: {errno_name} {str(e)}")
+    If ``max_depth`` is specified, does not search below that depth.
 
-    visited_dirs = set([_dir_id(stat_root)])
+    If ``globs`` is a list, files matching earlier entries are placed in the return value before
+    files matching later entries.
+    """
+    if not isinstance(globs, list):
+        globs = [globs]
+    # Apply normcase to file patterns and filenames to respect case insensitive filesystems.
+    regexes = [re.compile(fnmatch.translate(os.path.normcase(x))) for x in globs]
+    if not isinstance(roots, list):
+        roots = [roots]
+    # Ensure returned paths are always absolute.
+    roots = [os.path.abspath(r) for r in roots]
+    if max_depth is None:
+        max_depth = sys.maxsize
+    # Maps file pattern to list of matching paths, preserving order of the patterns.
+    found_files: Dict[re.Pattern, List[str]] = {pattern: [] for pattern in regexes}
+
+    # Each queue item stores the depth and path. This achieves a consistent traversal order by
+    # iterating through each directory in alphabetical order. This also traverses in BFS order to
+    # ensure finding the shortest path from any root to any file (in case of multiple shortest
+    # paths the one returned will be deterministic given the prior point).
+    dir_queue: Deque[Tuple[int, str]] = collections.deque()
+    visited_dirs: Set[Tuple[int, int]] = set()
+
+    # Enqueue unique roots. If any of the roots are not directories, raise an error. On any other
+    # error, log the issue and continue like we do during traversal too.
+    for root in roots:
+        try:
+            stat_root = os.stat(root)
+        except OSError as e:
+            _log_file_access_issue(e, root)
+            continue
+        dir_id = _dir_id(stat_root)
+        if dir_id not in visited_dirs:
+            dir_queue.appendleft((0, root))
+            visited_dirs.add(dir_id)
```

Comment:
Same input same output. [x, y] != [y, x]

Reply:
In a way we pass that message to the user through a debug message and we sometimes ask users to run spack -d to troubleshoot issues

Reply:
If we have

```
x/
  y/
  z/
```

and you do `find(roots=[x])`, I only want to fail if `x` is inaccessible, I don't want to fail if y/z are inaccessible.

I think there's a distinction between being unable to access the root, and being unable to access a child of the root.

> If you've ever run find you realize you never cared about permission errors, you always do find /usr ... 2>/dev/null

If I run `find /usr ... 2>/dev/null` and get 0 results, I usually get curious. `spack -d` is analogous to stripping off `2>/dev/null`, but I want to be as careful as we can be here, because the effects of missing a dir can take a long time to produce a failure: maybe one build succeeds and a later one fails (at which point one has no hope of backtracking through the debug output to the true culprit).

* I'd like to guard on the root dir not existing (in this analogy, misspelling the root of `find`), but like I said parts of our logic seem to depend on allowing that
* Detecting a permissions issue is more rare, but at least I can say that if the `/prefix/` of a package is inaccessible to us, that's a problem

Reply:
I don't find this argument convincing. 

> maybe one build succeeds and a later one fails

This is incredibly unlikely. If during a build `dep.libs` or `dep.headers` fails because a prefix of a dependency is not readable, the likelihood of the configure step failing immediately is *very* high.

> I'd like to guard on the root dir not existing 

This was not an error before, so not something to consider in this PR. (Before your PR, `os.walk` was used, which does not error. After your PR you specifically returned an empty list on ENOENT). Guarding against typos is not a consideration in this PR.

In the end I think what matters most is a simpler API for package.py where the average user doesn't need to handle errors twice. This is annoying:

```python
try:
    files = find([prefix.lib, prefix.lib64], ...)        
except (users have to look up what exception to catch):  # edge case 1
    files = []

if files:  # edge case 2
    do_something(files)
```

this is simpler:

```python
files = find([prefix.lib, prefix.lib64], ...)
if files:
    do_something(files)
```

Reply:
I think the convention we've used for stuff like this, e.g. `which()`, is a `required` argument. Currently the routine does not fail when it fails to find things, and the packages seem to expect that. In some cases they're looking for things like `pkgconfig` files and stuff like that, and an empty list is fine (as those files are not critical).

If we want this to fail on an empty list (where nothing is found) then I'd say add an optional `required=True` argument later (after this PR) to allow certain packages to use that.

Reply:
`required=True` is also poor API from the perspective of type checking, because the return type depends on a value (requires overloads + literals, and even then could be problematic if the value is forward to other function calls)

Reply:
I don't think this is the right call (and disagree with much of the reasoning), but since 3 people have approved this, I consider myself outvoted in this particular matter.